pkgs/native_assets_cli/lib/src/config.dart

```diff
@@ -0,0 +1,538 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert' hide json;
+import 'dart:io';
+
+import 'package:crypto/crypto.dart' show sha256;
+import 'package:pub_semver/pub_semver.dart';
+
+import 'api/deprecation_messages.dart';
+import 'architecture.dart';
+import 'build_mode.dart';
+import 'encoded_asset.dart';
+import 'json_utils.dart';
+import 'model/metadata.dart';
+import 'os.dart';
+import 'utils/datetime.dart';
+import 'utils/json.dart';
+
+/// The shared properties of a [LinkConfig] and a [BuildConfig].
+///
+/// This abstraction makes it easier to design APIs intended for both kinds of
+/// build hooks, building and linking.
+sealed class HookConfig {
+  /// The underlying json configuration of this [HookConfig].
+  final Map<String, Object?> json;
+
+  /// The version of the [HookConfig].
+  final Version version;
+
+  /// The directory in which output and intermediate artifacts that are unique
+  /// to this configuration can be placed.
+  ///
+  /// This directory is unique per hook and per configuration.
+  ///
+  /// The contents of this directory will not be modified by anything else than
+  /// the hook itself.
+  ///
+  /// The invoker of the the hook will ensure concurrent invocations wait on
+  /// each other.
+  final Uri outputDirectory;
+
+  /// The directory in which shared output and intermediate artifacts can be
+  /// placed.
+  ///
+  /// This directory is unique per hook.
+  ///
+  /// The contents of this directory will not be modified by anything else than
+  /// the hook itself.
+  ///
+  /// The invoker of the the hook will ensure concurrent invocations wait on
+  /// each other.
+  final Uri outputDirectoryShared;
+
+  /// The name of the package the assets are built for.
+  final String packageName;
+
+  /// The root of the package the assets are built for.
+  ///
+  /// Often a package's assets are built because a package is a dependency of
+  /// another. For this it is convenient to know the packageRoot.
+  final Uri packageRoot;
+
+  /// The operating system being compiled for.
+  final OS targetOS;
+
+  /// The [BuildMode] that the code should be compiled in.
+  ///
+  /// Currently [BuildMode.debug] and [BuildMode.release] are the only modes.
+  ///
+  /// Not available during a dry run.
+  final BuildMode? _buildMode;
+
+  /// The asset types that the invoker of this hook supports.
+  final List<String> supportedAssetTypes;
+
+  HookConfig(this.json)
```

Comment:
It looks like we lost validation and backwards compatibility.

Are you intending to remove support for protocol versions v1.4 and v1.5?

Reply:
We should gradually phase out support for older versions anyway - it's a pre-requisite to have a unique encoding / one version when we remove experiment flag. The sooner we do it the better - as it reduces maintenance burden.

=> So yes, it's intentional to remove old version support.

The stable branch cutoff (for 3.6.0) was a week ago, so anything we land now is going to be for the next stable.